src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,290 @@
+                                   const messenger::ZmqMessage &msg) {
+  switch (static_cast<MessageType>(msg.GetDestinationCallbackId())) {
+    case MessageType::ACK: {
+      nlohmann::json json = nlohmann::json::parse(msg.GetMessage());
+      REPLICATION_LOG_TRACE(fmt::format("ACK: {}", json.dump()));
+      auto callback_id = json.at("callback_id").get<uint64_t>();
+      auto callback = messenger->GetCallback(callback_id);
+      (*callback)(messenger, msg);
+      break;
+    }
+    case MessageType::HEARTBEAT:
+      REPLICATION_LOG_TRACE(fmt::format("Heartbeat from: {}", msg.GetRoutingId()));
+      break;
+    case MessageType::REPLICATE_BUFFER: {
+      // Parse the buffer from the received message and add it to the provided replication logs.
+      {
+        nlohmann::json message = nlohmann::json::parse(msg.GetMessage());
+        // TODO(WAN): Sanity-check the received message.
+
+        uint64_t source_callback_id = msg.GetSourceCallbackId();
+        uint64_t msg_id = message.at("buf_id");
+        REPLICATION_LOG_TRACE(fmt::format("ReplicateBuffer from: {} {}", msg.GetRoutingId(), msg_id));
+
+        // Check if the message needs to be buffered.
+        if (msg_id > last_record_received_id_ + 1) {
+          // The message should be buffered if there are gaps in between the last seen buffer.
+          received_message_queue_.push(msg);
+        } else {
+          // Otherwise, pull out the log record from the message and hand them to the replication log provider.
+          // TODO(WAN): And now we parse the same JSON repeatedly.. we could parse once and store <uint64,json> instead?
+          std::string content = nlohmann::json::from_cbor(message["content"].get<std::vector<uint8_t>>());
+          provider_->AddBufferFromMessage(source_callback_id, content);
+          last_record_received_id_ = msg_id;
+          // This may unleash the rest of the buffered messages.
+          while (!received_message_queue_.empty()) {
+            auto &temporary = received_message_queue_.top();
+            source_callback_id = temporary.GetSourceCallbackId();
+            message = nlohmann::json::parse(temporary.GetMessage());
+            msg_id = message.at("buf_id");
+            // Stop once you're missing a buffer.
+            if (msg_id > last_record_received_id_ + 1) {
+              break;
+            }
+            // Otherwise, send the top buffer's contents along.
+            received_message_queue_.pop();
+            NOISEPAGE_ASSERT(msg_id == last_record_received_id_ + 1, "Duplicate buffer? Old buffer?");
+            content = nlohmann::json::from_cbor(message["content"].get<std::vector<uint8_t>>());
+            provider_->AddBufferFromMessage(source_callback_id, content);
+            last_record_received_id_ = msg_id;
+          }
+        }
+      }
+      // TODO(WAN): Is it actually necessary to buffer the message first? A crashed replica is indistinguishable from
```

Comment:
Is this comment supposed to be here? It's a little unclear to me what it's referring to.

Reply:
that was a stale comment from when i was in crazytown until andy pointed out that doing an ACK here wouldn't be synchronous replication (since the replica wasn't really seeing the changes yet), will remove. we would want to ack somewhere early on here for async though.

Reply:
but we're not worrying about async for now